src/ripple/rpc/impl/Handler.cpp

```diff
@@ -148,45 +154,59 @@ class HandlerTable {
         return handlerTable;
     }
 
-    Handler const* getHandler(std::string name) const
+    Handler const* getHandler(unsigned version, std::string name) const
     {
-        auto i = table_.find(name);
-        return i == table_.end() ? nullptr : &i->second;
+        if(version > RPC::APIVersionSupportedRangeHigh || version < RPC::APIVersionSupportedRangeLow)
+            return nullptr;
+        auto & innerTable = table_[versionToIndex(version)];
+        auto i = innerTable.find(name);
+        return i == innerTable.end() ? nullptr : &i->second;
     }
 
     std::vector<char const*>
     getHandlerNames() const
     {
-        std::vector<char const*> ret;
-        ret.reserve(table_.size());
-        for (auto const& i : table_)
-            ret.push_back(i.second.name_);
-        return ret;
+        std::unordered_set<char const*> name_set;
+        for ( int index = 0; index < table_.size(); ++index)
+        {
+            for(auto const& h : table_[index])
+            {
+                name_set.insert(h.second.name_);
+            }
+        }
+        return std::vector<char const*>(name_set.begin(), name_set.end());
     }
 
   private:
-    std::map<std::string, Handler> table_;
+    std::array<std::map<std::string, Handler>, APINumberVersionSupported> table_;
 
     template <class HandlerImpl>
-    void addHandler()
+    void addHandler(unsigned version)
     {
-        assert (table_.find(HandlerImpl::name()) == table_.end());
+        assert (version >= RPC::APIVersionSupportedRangeLow && version <= RPC::APIVersionSupportedRangeHigh);
+        auto & innerTable = table_[versionToIndex(version)];
```

Comment:
I like "empty" table[0]. Probably not too much cost, but when we increase the value of APIVersionSupportedRangeLow, there will be more "empty" ones. Say APIVersionSupportedRangeLow==4, table[0], table[1], table[2], and table[3] will be empty.

Reply:
Ah, hmm... there's that, which I didn't think about. This may be a better fit for a map, but that comes at a minor performance cost.

I think this is fine as is, for now and let's see if anyone else has a better idea. Thanks.

Reply:
OK